### PR TITLE
@swm/add draggable list

### PR DESCRIPTION
**Pull request by kosmydel**
### Details

This PR adds support for both draggable lists on both web (using [react-beautiful-dnd](https://github.com/atlassian/react-beautiful-dnd) and native (using [react-native-draggable-flatlist](https://github.com/computerjazz/react-native-draggable-flatlist)).

### Related Issues

https://github.com/Expensify/App/issues/22716
https://github.com/Expensify/App/issues/25740

### CODE REVIEW

**Comment by kosmydel** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by kosmydel** [comment]
recheck

File: .prettierrrc.js
Comment (blazejkustra) [inline .prettierrrc.js:1]: Should be `.prettierrc.js`

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,20 @@
+MIT License
+
+Copyright (c) 2023 Jakub Kosmydel
```
Comment (blazejkustra) [inline LICENSE]: Change to Expensify

File: example/src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ * Prepare example items for the list
+ */
+const getItems = (count: number): Item[] =>
+  Array.from({ length: count }, (_, k) => k).map((k) => ({
+    id: `item-${k}`,
+    content: {
+      label: `Item #${k}`,
+      subtitle: `Subtitle ${k}`,
+      color: getBackgroundColor(k),
+    },
```
Comment (kosmydel) [inline example/src/App.tsx]: Use _.map()

File: example/src/utils.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+const COLORS = {
+  LIST: ['#FAF1E4', '#CEDEBD', '#9EB384'],
+  DARK: '#435334',
+  SUBTITLE: '#49573c',
+  ACTIVE: '#d6dbd0',
+  LIGHT: '#FAF1E4',
+};
+
+const getBackgroundColor = (index: number) => {
+  return COLORS.LIST[index % COLORS.LIST.length] ?? 'white';
```
Comment (kosmydel) [inline example/src/utils.tsx]: Don't use ??

File: src/DraggableList/index.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      onDragUpdate={onDragUpdate}
+    >
+      <Droppable droppableId="droppable">
+        {(droppableProvided) => (
+          <View
+            {...droppableProvided.droppableProps}
+            ref={droppableProvided.innerRef}
+          >
+            {items.map((item, index) => (
+              <Draggable key={item.id} draggableId={item.id} index={index}>
```
Comment (kosmydel) [inline src/DraggableList/index.tsx:77]: Add `keyExtractor` prop

File: src/DraggableList/index.native.tsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import type { DefaultItemProps, DraggableListProps } from './types';
+
+export const DraggableList = <T extends DefaultItemProps>({
+  items,
+  ...props
+}: DraggableListProps<T>) => {
+  return (
+    <DraggableFlatList
+      data={items}
+      keyExtractor={(item) => item.id}
```
Comment (kosmydel) [inline src/DraggableList/index.native.tsx:12]: Move this prop to the API

**Comment by hayata-suenaga** [comment]
@kosmydel Today, Expensify is internally decided not to create additional libraries if not needed.

How much functionality did you implement (or rather exposed) in this library? If we're just exposing just the functionality used in NewDot, then it might be better to move the code over to NewDot to make things easier.

Let me know what you think 🙇

**Comment by kosmydel** [comment]
> @kosmydel Today, Expensify is internally decided not to create additional libraries if not needed.
> 
> How much functionality did you implement (or rather exposed) in this library? If we're just exposing just the functionality used in NewDot, then it might be better to move the code over to NewDot to make things easier.
> 
> Let me know what you think 🙇

@hayata-suenaga 
I understand, here are my thoughts:

**My implementation:**
Since my approach was to use the `react-native-draggable-flatlist` API in the new library, on the native the code [was short](https://github.com/kosmydel/expensify-react-native-x-draggable-list/blob/a9afeb1009f926e44009cda3cfbcb78101c11566/src/DraggableList/index.native.tsx#L6-L9). The more code was implemented on the web side, to make the [react-beautiful-dnd](https://github.com/atlassian/react-beautiful-dnd) work with the native API. Because there were some differences, I had to adjust it a bit. I think that [the source](https://github.com/kosmydel/expensify-react-native-x-draggable-list/tree/%40swm/add-draggable-list/src) will tell more about how many things were implemented here.

I focused on exposing the functionality the NewDot app needed. There are also some extras like [decorators](https://github.com/kosmydel/expensify-react-native-x-draggable-list/tree/%40swm/add-draggable-list/src/decorators).

Of course, we can move this code to the NewDot app. Since this was mainly the wrapper, it might work. Please let me know about the next steps.

**Comment by hayata-suenaga** [comment]
Let's move the code over to App. I think we don't want to deal with the effort to maintain another library 🙇 

There are also some extras like [decorators](https://github.com/kosmydel/expensify-react-native-x-draggable-list/tree/%40swm/add-draggable-list/src/decorators).

could you tell me what decorators do? do we need them in App?

**Comment by kosmydel** [comment]
> Let's move the code over to App. I think we don't want to deal with the effort to maintain another library 🙇

Sure, I am going to prepare PR for that. As I have internal training until Friday, my availability will be limited.

> There are also some extras like [decorators](https://github.com/kosmydel/expensify-react-native-x-draggable-list/tree/%40swm/add-draggable-list/src/decorators).
> 
> could you tell me what decorators do? do we need them in App?

Those decorators work only on native, and 'decorate' items that are currently dragged (e.g. by scaling the item up a little bit or changing the opacity).
[Here is more context.](https://github.com/computerjazz/react-native-draggable-flatlist#cell-decorators) Those are probably not really important, but we might consider adding something like this. Moving this part of the code to the App should be easy.

**Comment by neil-marcellini** [comment]
Closing in favor of https://github.com/Expensify/App/pull/26307

**Pull request closed without merging**